### PR TITLE
feat: added polling jobs status system

### PR DESCRIPTION
**Pull request by CL-SHLOMIKONCHA**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✔                                                                        |
| Chore            | ✖                                                                       |

Add Job done logic:
* Call the callback URI’s instead of worker 
* when GPKG creation is finished - JSON serialize the raster “record“ object to json file named metadata.json in the same location
* Set Job success  // fail
* Helm
* Tests

### CODE REVIEW

File: helm/values.yaml [hunk 1]
```diff
@@ -73,6 +78,7 @@ env:
   logPrettyPrintEnabled: false
   responseCompressionEnabled: true
   requestPayloadLimit: 1mb
+  pollingTimeoutMS: 15000
```
Comment (asafMasa) [inline helm/values.yaml]: Needs to be smaller (2 sec )

File: helm/values.yaml [hunk 2]
```diff
@@ -97,7 +103,7 @@ resources:
       memory: 128Mi
 
 route:
-  enabled: true
+  enabled: false
```
Comment (asafMasa) [inline helm/values.yaml]: why?

File: src/clients/callbackClient.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+import { ICallbackData, IConfig } from '../common/interfaces';
+
+@singleton()
+export class CallbackClient extends HttpClient {
+  public constructor(@inject(SERVICES.LOGGER) logger: Logger, @inject(SERVICES.CONFIG) private readonly config: IConfig) {
+    super(logger, '', 'requestCallback', config.get<IHttpRetryConfig>('httpRetry'));
+  }
+
+  public async send(callbackUrl: string, params: ICallbackData): Promise<void> {
+    this.logger.info(`send Callback request to URL: ${callbackUrl} with data ${JSON.stringify(params)}`);
```
Comment (asafMasa) [inline src/clients/callbackClient.ts:14]: try to use these format
this.logger.info(params, `send Callback request to URL: ${callbackUrl} with data`);

File: src/common/enums.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+//TODO: replace with model
```
Comment (asafMasa) [inline src/common/enums.ts]: delete and use from [mc-priority-queue](https://github.com/MapColonies/mc-priority-queue)

File: src/common/interfaces.ts [hunk 5]
```diff
@@ -85,7 +87,8 @@ export interface IJobParameters {
   callbacks: ICallbackTarget[];
   sanitizedBbox: BBox2d;
   zoomLevel: number;
-  callbackParams?: ICallbackParams;
+  callbackParams?: ICallbackDataBase;
+  fileName?: string;
```
Comment (asafMasa) [inline src/common/interfaces.ts]: why the fileName is nullable?

File: src/common/utils.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+export const generatePackageName = (dbId: string, zoomLevel: number, bbox: BBox): string => {
+  const numberOfDecimals = 5;
+  const bboxToString = bbox.map((val) => String(val.toFixed(numberOfDecimals)).replace('.', '_').replace(/-/g, 'm')).join('');
+  return `gm_${dbId.replace(/-/g, '_')}_${zoomLevel}_${bboxToString}.gpkg`;
+};
+
+export const getGpkgFilePath = (gpkgsLocation: string, packageName: string): string => {
+  const packageDirectoryName = packageName.substr(0, packageName.lastIndexOf('.'));
+  const packageFullPath = join(gpkgsLocation, packageDirectoryName, packageName);
```
Comment (asafMasa) [inline src/common/utils.ts:24]: check if compatible for windows look at https://github.com/MapColonies/exporter-trigger/pull/31/files#diff-db223db1ee3a87797b8896c9bfdb421cca1fef09dabfe4499eacafae73db5d7eR127

File: src/createPackage/models/createPackageManager.ts [hunk 7]
```diff
@@ -109,6 +116,14 @@ export class CreatePackageManager {
     return duplicationExist;
   }
 
+  public async createJsonMetadata(filePath: string, dbId: string): Promise<void> {
+    const fileName = 'metadata.json';
+    const metadataFilePath = join(dirname(filePath), fileName);
```
Comment (asafMasa) [inline src/createPackage/models/createPackageManager.ts]: pay attention to getSeparator method - https://github.com/MapColonies/exporter-trigger/pull/31/files#diff-db223db1ee3a87797b8896c9bfdb421cca1fef09dabfe4499eacafae73db5d7eR127

File: src/createPackage/models/createPackageManager.ts [hunk 8]
```diff
@@ -71,9 +75,11 @@ export class CreatePackageManager {
         batches.push(bboxToTileRange(sanitizedBbox, i));
       }
       const separator = this.getSeparator();
+      const packageName = generatePackageName(dbId, zoomLevel, sanitizedBbox);
+      const packageFullPath = getGpkgFilePath(this.gpkgsLocation, packageName);
       const sources: IMapSource[] = [
         {
-          path: this.generatePackageName(dbId, zoomLevel, sanitizedBbox), //gpkg path
+          path: packageFullPath,
```
Comment (asafMasa) [inline src/createPackage/models/createPackageManager.ts]: check tiles-Merger expected input, full or relative?

File: src/createPackage/models/createPackageManager.ts [hunk 9]
```diff
@@ -109,6 +116,14 @@ export class CreatePackageManager {
     return duplicationExist;
   }
 
+  public async createJsonMetadata(filePath: string, dbId: string): Promise<void> {
+    const fileName = 'metadata.json';
```
Comment (asafMasa) [inline src/createPackage/models/createPackageManager.ts]: should be initialized once

File: src/tasks/models/tasksManager.ts [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+  public async finalizeJob(job: JobResponse, expirationDate: Date, isSuccess = true, reason?: string): Promise<void> {
+    let updateJobParams: IUpdateJobBody<IJobParameters> = {
+      status: isSuccess ? OperationStatus.COMPLETED : OperationStatus.FAILED,
+      reason,
+      percentage: isSuccess ? 100 : undefined,
+      expirationDate: expirationDate,
+    };
+    try {
+      if (isSuccess) {
+        const packageName = job.parameters.fileName as string;
```
Comment (asafMasa) [inline src/tasks/models/tasksManager.ts]: add log

File: src/tasks/models/tasksManager.ts [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    const completedJobs = jobs?.filter((job) => job.completedTasks === job.taskCount);
+    const failedJobs = jobs?.filter((job) => job.failedTasks === job.taskCount);
+    const jobsStatus = {
+      completed: completedJobs,
+      failed: failedJobs,
+    };
+    return jobsStatus;
+  }
+
+  // public async getJobsByCompletedTasks(): Promise<JobResponse[]> {
```
Comment (asafMasa) [inline src/tasks/models/tasksManager.ts]: remove remarks

File: src/index.ts [hunk 12]
```diff
@@ -22,7 +23,24 @@ const app = getApp();
...
 const server = createTerminus(createServer(app), { healthChecks: { '/liveness': stubHealthcheck, onSignal: container.resolve('onSignal') } });
-
+const pollingManager: PollingManager = container.resolve(POLLING_MANGER_SYMBOL);
 server.listen(port, () => {
   logger.info(`app started on port ${port}`);
 });
+const mainPollLoop = async (): Promise<void> => {
+  const pollingTimout = config.get<number>('pollingTimeoutMS');
+  const isRunning = true;
+  logger.debug('running job status poll');
```
Comment (asafMasa) [inline src/index.ts]: info

File: src/pollingManager.ts [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+@singleton()
+export class PollingManager {
+  private readonly expirationDate: number;
+  public constructor(@inject(SERVICES.LOGGER) private readonly logger: Logger, @inject(TasksManager) private readonly taskManager: TasksManager) {
+    this.expirationDate = config.get<number>('jobManager.expirationTime');
+  }
+
+  public async jobStatusPoll(): Promise<void> {
+    const jobs = await this.taskManager.getJobsByTaskStatus();
+    console.log(jobs);
```
Comment (asafMasa) [inline src/pollingManager.ts]: remove console

File: src/index.ts [hunk 14]
```diff
@@ -22,7 +23,24 @@ const app = getApp();
...
+  //eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+  while (isRunning) {
+    try {
+      await pollingManager.jobStatusPoll();
+    } catch (error) {
+      logger.error(`mainPollLoop: Error: ${JSON.stringify(error, Object.getOwnPropertyNames(error))}`);
+    } finally {
+      await new Promise((resolve) => setTimeout(resolve, pollingTimout));
```
Comment (asafMasa) [inline src/index.ts]: wait only if no data was received from job manager

File: src/tasks/models/tasksManager.ts [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+  //   const jobs = await this.jobManagerClient.getJobsStatus();
+  //   const completedJobs = jobs?.filter((job) => (job.completedTasks === job.taskCount) || (job.failedTasks === job.taskCount));
+  //   return completedJobs!;
+  // }
+
+  public async getTaskStatusByJobId(jobId: string): Promise<ITaskStatusResponse> {
+    this.logger.info(`Getting task status by jobId: ${jobId}`);
```
Comment (asafMasa) [inline src/tasks/models/tasksManager.ts]: why do you need this log?, if you really need it use debug

File: src/tasks/models/tasksManager.ts [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+    const tasks = await this.jobManagerClient.getTasksByJobId(jobId);
+
+    if (tasks.length === 0) {
+      throw new NotFoundError(`jobId: ${jobId} is not exists`);
```
Comment (asafMasa) [inline src/tasks/models/tasksManager.ts]: No tasks where found for jobId: ${jobId}

File: src/tasks/models/tasksManager.ts [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+    const statusResponse: ITaskStatusResponse = {
+      percentage: task.percentage,
+      status: task.status,
+    };
+    return statusResponse;
+  }
+
+  public async sendCallbacks(job: JobResponse, expirationDate: Date, errorReason?: string): Promise<ICallbackDataBase | undefined> {
+    try {
+      const downloadServerUrl = config.get('downloadServerUrl');
```
Comment (asafMasa) [inline src/tasks/models/tasksManager.ts]: get once (ctor?)

File: src/tasks/models/tasksManager.ts [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+      const packageName = job.parameters.fileName as string;
```
Comment (asafMasa) [inline src/tasks/models/tasksManager.ts]: does it includes the directory?

File: src/tasks/models/tasksManager.ts [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+    this.gpkgsLocation = config.get<string>('gpkgsLocation');
+    this.expirationDate = config.get<number>('jobManager.expirationTime');
+  }
+
+  public async getJobsByTaskStatus(): Promise<IJobStatusResponse> {
+    const jobs = await this.jobManagerClient.getJobsStatus();
+    const completedJobs = jobs?.filter((job) => job.completedTasks === job.taskCount);
+    const failedJobs = jobs?.filter((job) => job.failedTasks === job.taskCount);
+    const jobsStatus = {
+      completed: completedJobs,
```
Comment (asafMasa) [inline src/tasks/models/tasksManager.ts]: completed rename to -> CompletedJobs

File: src/tasks/models/tasksManager.ts [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+      failed: failedJobs,
```
Comment (asafMasa) [inline src/tasks/models/tasksManager.ts]: failedJobs

File: src/pollingManager.ts [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+    const expirationDate = new Date();
+    expirationDate.setDate(expirationDate.getDate() + this.expirationDate);
+    if (jobs.completed && jobs.completed.length) {
```
Comment (asafMasa) [inline src/pollingManager.ts]: jobs.completed.length >0

File: src/pollingManager.ts [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+      this.logger.debug(`completed jobs detected, running finalize job`);
```
Comment (asafMasa) [inline src/pollingManager.ts]: info

File: src/pollingManager.ts [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+      jobs.completed.forEach((job) => {
+        this.taskManager.finalizeJob(job, expirationDate);
+      });
+    } else if (jobs.failed && jobs.failed.length) {
+      this.logger.debug(`failed jobs detected, updating job status`);
```
Comment (asafMasa) [inline src/pollingManager.ts]: info

File: src/tasks/interfaces.ts [hunk 24]
```diff
@@ -0,0 +1,32 @@
+import { LayerMetadata } from '@map-colonies/mc-model-types';
```
Comment (asafMasa) [inline src/tasks/interfaces.ts]: remove file

File: src/tasks/models/tasksManager.ts (hunk 17 above)
Comment (asafMasa) [inline src/tasks/models/tasksManager.ts]: add send callbacks log

File: src/pollingManager.ts [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+      jobs.failed.forEach((job) => {
+        this.taskManager.finalizeJob(job, expirationDate, false);
```
Comment (asafMasa) [inline src/pollingManager.ts]: add reason -> failed to create GPKG

File: src/tasks/models/tasksManager.ts [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+
+  public async finalizeJob(job: JobResponse, expirationDate: Date, isSuccess = true, reason?: string): Promise<void> {
+    let updateJobParams: IUpdateJobBody<IJobParameters> = {
+      status: isSuccess ? OperationStatus.COMPLETED : OperationStatus.FAILED,
+      reason,
+      percentage: isSuccess ? 100 : undefined,
+      expirationDate: expirationDate,
+    };
+    try {
+      if (isSuccess) {
```
Comment (asafMasa) [inline src/tasks/models/tasksManager.ts:116]: callbacks needs to be called also on failure

**Review by asafMasa — CHANGES_REQUESTED** [review]
👍

File: src/common/interfaces.ts [hunk 27]
```diff
@@ -1,4 +1,4 @@
-import { BBox2d } from '@turf/helpers/dist/js/lib/geojson';
+import { BBox, BBox2d, MultiPolygon, Polygon } from '@turf/helpers/dist/js/lib/geojson';
```
Comment (cl-shaharshavit-rnd) [inline src/common/interfaces.ts]: split import to import everything apart from bbox2d from @turf/turf

File: src/common/utils.ts [hunk 28]
```diff
@@ -0,0 +1,20 @@
+import { promises as fsPromise } from 'fs';
+import { join } from 'path';
+import { BBox } from '@turf/helpers/dist/js/lib/geojson';
```
Comment (cl-shaharshavit-rnd) [inline src/common/utils.ts]: import from @turf/turf or @turf/bbox 
instead

File: src/common/utils.ts (hunk 6 above)
Comment (cl-shaharshavit-rnd) [reply src/common/utils.ts:24]: this is internal path and join is compatible with windows

File: src/createPackage/models/createPackageManager.ts (hunk 8 above)
Comment (cl-shaharshavit-rnd) [reply src/createPackage/models/createPackageManager.ts]: it if doesn't expect relative its bug in the merger

File: src/createPackage/models/createPackageManager.ts (hunk 9 above)
Comment (cl-shaharshavit-rnd) [reply src/createPackage/models/createPackageManager.ts]: ? 
it isn't being used in other places and it is irrelevant in performance context

File: src/createPackage/models/createPackageManager.ts (hunk 7 above)
Comment (cl-shaharshavit-rnd) [reply src/createPackage/models/createPackageManager.ts]: this is better when not working with s3

File: src/pollingManager.ts [hunk 29]
```diff
@@ -0,0 +1,33 @@
...
+import { Logger } from '@map-colonies/js-logger';
+import { SERVICES } from './common/constants';
+import { TasksManager } from './tasks/models/tasksManager';
+
+export const POLLING_MANGER_SYMBOL = Symbol('tasksFactory');
+
+@singleton()
+export class PollingManager {
+  private readonly expirationDate: number;
+  public constructor(@inject(SERVICES.LOGGER) private readonly logger: Logger, @inject(TasksManager) private readonly taskManager: TasksManager) {
```
Comment (cl-shaharshavit-rnd) [inline src/pollingManager.ts:14]: add line break before ctor

File: src/pollingManager.ts [hunk 30]
```diff
@@ -0,0 +1,33 @@
...
+import { inject, singleton } from 'tsyringe';
+import { Logger } from '@map-colonies/js-logger';
+import { SERVICES } from './common/constants';
+import { TasksManager } from './tasks/models/tasksManager';
+
+export const POLLING_MANGER_SYMBOL = Symbol('tasksFactory');
+
+@singleton()
+export class PollingManager {
+  private readonly expirationDate: number;
```
Comment (cl-shaharshavit-rnd) [inline src/pollingManager.ts]: this is bad name as it is more of ttl (in what units?) then date

File: src/pollingManager.ts (hunk 21 above)
Comment (cl-shaharshavit-rnd) [reply src/pollingManager.ts]: jobs.completed?.length >0

File: src/pollingManager.ts [hunk 31]
```diff
@@ -0,0 +1,33 @@
...
+
+  public async jobStatusPoll(): Promise<void> {
+    const jobs = await this.taskManager.getJobsByTaskStatus();
+    console.log(jobs);
+    const expirationDate = new Date();
+    expirationDate.setDate(expirationDate.getDate() + this.expirationDate);
+    if (jobs.completed && jobs.completed.length) {
+      this.logger.debug(`completed jobs detected, running finalize job`);
+      jobs.completed.forEach((job) => {
+        this.taskManager.finalizeJob(job, expirationDate);
```
Comment (cl-shaharshavit-rnd) [inline src/pollingManager.ts]: lint

File: src/pollingManager.ts [hunk 32]
```diff
@@ -0,0 +1,33 @@
...
+      });
+    } else if (jobs.failed && jobs.failed.length) {
```
Comment (cl-shaharshavit-rnd) [inline src/pollingManager.ts]: lint

File: src/pollingManager.ts (hunk 25 above)
Comment (cl-shaharshavit-rnd) [inline src/pollingManager.ts]: lint

File: src/tasks/models/tasksManager.ts [hunk 33]
```diff
@@ -0,0 +1,130 @@
...
+      status: task.status,
+    };
+    return statusResponse;
+  }
+
+  public async sendCallbacks(job: JobResponse, expirationDate: Date, errorReason?: string): Promise<ICallbackDataBase | undefined> {
+    try {
+      const downloadServerUrl = config.get('downloadServerUrl');
+      const packageName = job.parameters.fileName as string;
+      const fileUri = `${downloadServerUrl}/downloads/${packageName}`;
```
Comment (cl-shaharshavit-rnd) [inline src/tasks/models/tasksManager.ts]: lint

File: src/tasks/models/tasksManager.ts [hunk 34]
```diff
@@ -0,0 +1,130 @@
...
+    } catch (error) {
+      this.logger.error(`Sending callbacks has failed with error: ${JSON.stringify(error, Object.getOwnPropertyNames(error))}`);
+    }
+  }
+
+  public async finalizeJob(job: JobResponse, expirationDate: Date, isSuccess = true, reason?: string): Promise<void> {
+    let updateJobParams: IUpdateJobBody<IJobParameters> = {
+      status: isSuccess ? OperationStatus.COMPLETED : OperationStatus.FAILED,
+      reason,
+      percentage: isSuccess ? 100 : undefined,
```
Comment (cl-shaharshavit-rnd) [inline src/tasks/models/tasksManager.ts:110]: lint (ignore)

File: src/tasks/models/tasksManager.ts [hunk 35]
```diff
@@ -0,0 +1,130 @@
...
+        const packageName = job.parameters.fileName as string;
+        const packageFullPath = getGpkgFilePath(this.gpkgsLocation, packageName);
+        await this.packageManager.createJsonMetadata(packageFullPath.substr(0, packageFullPath.lastIndexOf('.')), job.internalId as string);
+        const callbackParams = await this.sendCallbacks(job, expirationDate);
+        updateJobParams = { ...updateJobParams, parameters: { ...job.parameters, callbackParams } };
+      }
+      this.logger.info(`Update Job status to success=${String(isSuccess)} jobId=${job.id}`);
+      await this.jobManagerClient.updateJob(job.id, updateJobParams);
+    } catch (error) {
+      this.logger.error(`Could not finalize job: ${job.id} updating failed job status, error: ${error}`);
```
Comment (cl-shaharshavit-rnd) [inline src/tasks/models/tasksManager.ts]: lint

**Review by cl-shaharshavit-rnd — CHANGES_REQUESTED** [review]

File: helm/values.yaml (hunk 1 above)
Comment (CL-SHLOMIKONCHA) [reply helm/values.yaml]: fixed

File: helm/values.yaml (hunk 2 above)
Comment (CL-SHLOMIKONCHA) [reply helm/values.yaml]: minikube local check - fixed

File: src/common/enums.ts (hunk 4 above)
Comment (CL-SHLOMIKONCHA) [reply src/common/enums.ts]: fixed

File: src/pollingManager.ts (hunk 23 above)
Comment (CL-SHLOMIKONCHA) [reply src/pollingManager.ts]: fixed

File: src/tasks/interfaces.ts (hunk 24 above)
Comment (CL-SHLOMIKONCHA) [reply src/tasks/interfaces.ts]: removed

File: src/tasks/models/tasksManager.ts (hunk 17 above)
Comment (CL-SHLOMIKONCHA) [reply src/tasks/models/tasksManager.ts]: fixed

File: src/pollingManager.ts (hunk 25 above)
Comment (CL-SHLOMIKONCHA) [reply src/pollingManager.ts]: fixed

File: src/tasks/models/tasksManager.ts (hunk 26 above)
Comment (CL-SHLOMIKONCHA) [reply src/tasks/models/tasksManager.ts:116]: fixed

File: src/common/interfaces.ts (hunk 27 above)
Comment (CL-SHLOMIKONCHA) [reply src/common/interfaces.ts]: fixed

File: src/common/utils.ts (hunk 28 above)
Comment (CL-SHLOMIKONCHA) [reply src/common/utils.ts]: fixed

File: src/pollingManager.ts (hunk 29 above)
Comment (CL-SHLOMIKONCHA) [reply src/pollingManager.ts:14]: fixed

File: src/tasks/models/tasksManager.ts (hunk 11 above)
Comment (CL-SHLOMIKONCHA) [reply src/tasks/models/tasksManager.ts]: removed

File: src/tasks/models/tasksManager.ts (hunk 15 above)
Comment (CL-SHLOMIKONCHA) [reply src/tasks/models/tasksManager.ts]: removed

File: src/tasks/models/tasksManager.ts (hunk 16 above)
Comment (CL-SHLOMIKONCHA) [reply src/tasks/models/tasksManager.ts]: fixed

File: src/tasks/models/tasksManager.ts (hunk 20 above)
Comment (CL-SHLOMIKONCHA) [reply src/tasks/models/tasksManager.ts]: fixed

File: src/tasks/models/tasksManager.ts (hunk 19 above)
Comment (CL-SHLOMIKONCHA) [reply src/tasks/models/tasksManager.ts]: fixed

File: src/pollingManager.ts (hunk 21 above)
Comment (CL-SHLOMIKONCHA) [reply src/pollingManager.ts]: fixed

File: src/pollingManager.ts (hunk 22 above)
Comment (CL-SHLOMIKONCHA) [reply src/pollingManager.ts]: fixed

File: src/pollingManager.ts (hunk 30 above)
Comment (CL-SHLOMIKONCHA) [reply src/pollingManager.ts]: fixed

File: src/pollingManager.ts (hunk 13 above)
Comment (CL-SHLOMIKONCHA) [reply src/pollingManager.ts]: fixed

File: src/index.ts (hunk 12 above)
Comment (CL-SHLOMIKONCHA) [reply src/index.ts]: as you wanted it every 2 sec i wont recommend for info log, going to spam..

File: src/tasks/models/tasksManager.ts (hunk 10 above)
Comment (CL-SHLOMIKONCHA) [reply src/tasks/models/tasksManager.ts]: fixed

File: src/common/interfaces.ts (hunk 5 above)
Comment (CL-SHLOMIKONCHA) [reply src/common/interfaces.ts]: fixed

File: src/clients/callbackClient.ts (hunk 3 above)
Comment (CL-SHLOMIKONCHA) [reply src/clients/callbackClient.ts:14]: ?

File: src/createPackage/models/createPackageManager.ts (hunk 9 above)
Comment (CL-SHLOMIKONCHA) [reply src/createPackage/models/createPackageManager.ts]: fixed

File: src/tasks/models/tasksManager.ts (hunk 33 above)
Comment (CL-SHLOMIKONCHA) [reply src/tasks/models/tasksManager.ts]: fixed

File: src/tasks/models/tasksManager.ts (hunk 35 above)
Comment (CL-SHLOMIKONCHA) [reply src/tasks/models/tasksManager.ts]: fixed

File: src/tasks/models/tasksManager.ts (hunk 34 above)
Comment (CL-SHLOMIKONCHA) [reply src/tasks/models/tasksManager.ts:110]: fixed

File: src/pollingManager.ts (hunk 25 above)
Comment (CL-SHLOMIKONCHA) [reply src/pollingManager.ts]: fixed

File: src/pollingManager.ts (hunk 32 above)
Comment (CL-SHLOMIKONCHA) [reply src/pollingManager.ts]: fixed

File: src/pollingManager.ts (hunk 31 above)
Comment (CL-SHLOMIKONCHA) [reply src/pollingManager.ts]: fixed

File: src/tasks/models/tasksManager.ts (hunk 17 above)
Comment (CL-SHLOMIKONCHA) [reply src/tasks/models/tasksManager.ts]: fixed

File: src/createPackage/models/createPackageManager.ts (hunk 8 above)
Comment (CL-SHLOMIKONCHA) [reply src/createPackage/models/createPackageManager.ts]: fixed

File: src/index.ts (hunk 14 above)
Comment (CL-SHLOMIKONCHA) [reply src/index.ts]: fixed

File: src/tasks/models/tasksManager.ts (hunk 18 above)
Comment (CL-SHLOMIKONCHA) [reply src/tasks/models/tasksManager.ts]: fixed

File: helm/values.yaml [hunk 36]
```diff
@@ -73,14 +78,15 @@ env:
   logPrettyPrintEnabled: false
   responseCompressionEnabled: true
   requestPayloadLimit: 1mb
+  pollingTimeoutMS: 15000
```
Comment (asafMasa) [inline helm/values.yaml]: 2000 default

File: src/clients/callbackClient.ts (hunk 3 above)
Comment (asafMasa) [reply src/clients/callbackClient.ts:14]: talk with @ronenkapelian or with me

File: src/tasks/models/tasksManager.ts [hunk 37]
```diff
@@ -0,0 +1,127 @@
...
+    };
+    return statusResponse;
+  }
+
+  public async sendCallbacks(job: JobResponse, expirationDate: Date, errorReason?: string): Promise<ICallbackDataBase | undefined> {
+    try {
+      this.logger.info(`Sending callback for job: ${job.id}`);
+      const packageName = job.parameters.fileName;
+      const fileRelativePath = getGpkgRelativePath(packageName);
+      const fileUri = `${this.downloadServerUrl}/downloads/${fileRelativePath}`;
```
Comment (asafMasa) [inline src/tasks/models/tasksManager.ts]: why do you calculate "packageFullPath" or put URI if success == false?

**Review by asafMasa — CHANGES_REQUESTED** [review]
Good Job!

File: src/tasks/models/tasksManager.ts (hunk 37 above)
Comment (asafMasa) [reply src/tasks/models/tasksManager.ts]: Fix tests

File: helm/values.yaml (hunk 36 above)
Comment (CL-SHLOMIKONCHA) [reply helm/values.yaml]: fixed

File: src/tasks/models/tasksManager.ts (hunk 37 above)
Comment (CL-SHLOMIKONCHA) [reply src/tasks/models/tasksManager.ts]: URI fixed

File: src/clients/callbackClient.ts (hunk 3 above)
Comment (CL-SHLOMIKONCHA) [reply src/clients/callbackClient.ts:14]: this service is using old version of 'js-logger' (0.0.3) current boiler plate version 'js-logger' is 0.0.5, we need create new issue about update logger version and fix those all logs in the service

File: src/createPackage/models/createPackageManager.ts [hunk 38]
```diff
@@ -179,7 +178,7 @@ export class CreatePackageManager {
 
       return {
         id: processingJob.id,
-        taskIds: processingJob.tasks.map((t) => t.id),
+        taskIds: processingJob.tasks!.map((t) => t.id),
```
Comment (cl-shaharshavit-rnd) [inline src/createPackage/models/createPackageManager.ts]: lint

File: tests/unit/createPackage/models/createPackageModel.spec.ts [hunk 39]
```diff
@@ -39,126 +29,126 @@ describe('CreatePackageManager', () => {
   });
 
   describe('#create', () => {
-    it('should create job and return its job and task ids', async () => {
+    it('should create job and return his job and task ids', async () => {
```
Comment (cl-shaharshavit-rnd) [inline tests/unit/createPackage/models/createPackageModel.spec.ts]: job is it

File: tests/unit/createPackage/models/createPackageModel.spec.ts [hunk 40]
```diff
@@ -39,126 +29,126 @@ describe('CreatePackageManager', () => {
...
 
-      // eslint-disable-next-line
-      const checkForDuplicateResponse = await (createPackageManager as unknown as { checkForDuplicate: any }).checkForDuplicate(jobDupParams, []);
-
-      await createPackageManager.createPackage(userInput);
+      const res = await createPackageManager.createPackage(userInput);
       const expectedReturn: ICreateJobResponse = {
         id: inProgressJob.id,
-        taskIds: [inProgressJob.tasks[0].id],
+        taskIds: [inProgressJob.tasks![0].id],
```
Comment (cl-shaharshavit-rnd) [inline tests/unit/createPackage/models/createPackageModel.spec.ts]: lint

**Review by cl-shaharshavit-rnd — CHANGES_REQUESTED** [review]
great work on tests 

FIX LINT!!!

File: src/createPackage/models/createPackageManager.ts (hunk 38 above)
Comment (CL-SHLOMIKONCHA) [reply src/createPackage/models/createPackageManager.ts]: fixed

File: tests/unit/createPackage/models/createPackageModel.spec.ts (hunk 39 above)
Comment (CL-SHLOMIKONCHA) [reply tests/unit/createPackage/models/createPackageModel.spec.ts]: fixed

File: tests/unit/createPackage/models/createPackageModel.spec.ts (hunk 40 above)
Comment (CL-SHLOMIKONCHA) [reply tests/unit/createPackage/models/createPackageModel.spec.ts]: fixed

**Review by cl-shaharshavit-rnd — APPROVED** [review]

**Review by asafMasa — APPROVED** [review]

**Pull request merged**